lib/token/symmetric.go

```diff
@@ -0,0 +1,194 @@
+	if err != nil {
+		return nil, err
+	}
+	if n != int(size/8) {
+		return nil, fmt.Errorf("PRNG could not provide %d bytes of key", size)
+	}
+
+	return key, nil
+}
+
+// Creates a new random key and stores it in settings, or return error.
+//
+// size represents the key size in bits. If size is 0, uses a default key size
+// of 256 bits.
+func WithGeneratedSymmetricKey(size int) SymmetricSetter {
+	return func(be *SymmetricEncoder) error {
+		key, err := GenerateSymmetricKey(be.rng, size)
+		if err != nil {
+			return err
+		}
+		return UseSymmetricKey(key)(be)
+	}
+}
+
+// Reads a key from a file, or creates a new one and stores it in a file.
+// Returns error if it can't succeed in generating or storing a new key.
+//
+// size is the size in bits of the desired key. If left to 0, defaults to
+// 256 bits.
+//
+// The generated (or read) file is just the raw content of the key. For
+// example, for a key of 256 bits, it will generate a file of exactly
+// 32 bytes, containing the binary encoded key.
+func ReadOrGenerateSymmetricKey(path string, size int) SymmetricSetter {
+	if size == 0 {
+		size = 256
+	}
+
+	return func(be *SymmetricEncoder) error {
+		var err error
+		var key []byte
+		if path != "" {
+			key, err = ioutil.ReadFile(path)
+		}
+		if err != nil || len(key) <= 0 || len(key)*8 != size {
+			if err := WithGeneratedSymmetricKey(size)(be); err != nil {
+				return err
+			}
+			if path != "" {
+				err = ioutil.WriteFile(path, be.key, 0600)
+			}
+		} else {
+			err = UseSymmetricKey(key)(be)
+		}
+		return err
+	}
+}
+
+// NewSymmetricEncoder creates a new encoder using AES in GCM mode as a symmetric cipher.
+//
+// A typical way to use the encoder would be:
+//
+//    rng := rand.New(srand.Source)  // using github.com/enkit/lib/srand library.
+//    ...
+//    be, err := NewSymmetricEncoder(rng, ReadOrGenerateSymmetricKey("/etc/keys/connect.key", 0))
+//    if err != nil ...
+//
+// Or:
+//
+//    key, err := GenerateSymmetricKey(rng, 0)
+//    if err != nil ...
+//
+//    be, err := NewSymmetricEncoder(rng, UseSymmetricKey(key))
+//    ...
+//
+// followed by calls to Encode() and Decode().
+func NewSymmetricEncoder(rng *rand.Rand, setters ...SymmetricSetter) (*SymmetricEncoder, error) {
+	be := &SymmetricEncoder{rng: rng}
+	for _, setter := range setters {
+		if err := setter(be); err != nil {
+			return nil, err
+		}
+	}
+
+	if len(be.key) <= 0 {
+		return nil, fmt.Errorf("No key set")
+	}
+
+	block, err := aes.NewCipher(be.key)
+	if err != nil {
+		return nil, err
+	}
+
+	be.cipher, err = cipher.NewGCM(block)
+	if err != nil {
+		return nil, err
+	}
+
+	return be, nil
+}
+
+// SymmetricCreator is a CryptoFactory suitable for use with NewMultiKeyCryptoEncoder().
+var SymmetricCreator CryptoFactory = func(rng *rand.Rand, key []byte) (BinaryEncoder, []byte, error) {
+	if key != nil {
+		be, err := NewSymmetricEncoder(rng, UseSymmetricKey(key))
+		return be, key, err
+	}
+
+	key, err := GenerateSymmetricKey(rng, 0)
```

Comment:
I don't like this behaviour. I feel like it shouldn't default to random if not set. I don't know feels like it could lead to a lot of misleading tests for applications.

Reply:
In this specific function, CryptoFactory, it has to be a random key. This is used by the MultiCrypto encoder, which needs "something" to a) provide a random/unguessable/secret key to use to encrypt the message once, and then 2) an encoder capable of encrypting that data.

Then, this unguessable key is encrypted with each key that should have access to the message in turn, so any of those *other* keys can be used for crypto purposes.

The other encoders default to use a random key as a) it makes development simpler (start server, it works), b) it is the safest (eg, no secret stored on disk), and c) is "acceptable" in many deployments (single backend, restarted rarely, tokens are valid until timeout or until backend is restarted).

But also orthogonal to this PR?